python/ray/autoscaler/v2/instance_manager/node_provider.py

```diff
@@ -1,16 +1,160 @@
 import logging
-from abc import ABCMeta, abstractmethod
-from typing import Dict, List, Set
+from abc import ABC, ABCMeta, abstractmethod
+from dataclasses import dataclass, field
+from typing import Dict, List, Optional, Set
 
 from ray.autoscaler._private.node_launcher import BaseNodeLauncher
 from ray.autoscaler.node_provider import NodeProvider as NodeProviderV1
 from ray.autoscaler.tags import TAG_RAY_USER_NODE_TYPE
+from ray.autoscaler.v2.schema import NodeType
 from ray.autoscaler.v2.instance_manager.config import AutoscalingConfig
 from ray.core.generated.instance_manager_pb2 import Instance
 
 logger = logging.getLogger(__name__)
 
 
+# Type Alias. This is a **unique identifier** for a cloud node in the cluster.
+# The node provider should guarantee that this id is unique across the cluster,
+# such that:
+#   - When a cloud node is created and running, no other cloud node in the
+#     cluster has the same id.
+#   - When a cloud node is terminated, no other cloud node in the cluster should
+#     be assigned the same id later.
+CloudInstanceId = str
+
+
+@dataclass
+class CloudInstance:
+    # The cloud instance id.
+    cloud_instance_id: CloudInstanceId
+    # The node type of the cloud instance.
+    node_type: NodeType
+
+
+@dataclass
+class CloudNodeProviderError:
+    """
+    An error class that represents an error that happened in the cloud node provider.
+    """
+
+    # The exception that caused the error.
+    exception: Optional[Exception] = None
+    # The details of the error.
+    details: Optional[str] = None
+    # The timestamp of the error in nanoseconds.
+    timestamp_ns: int
+
+
+@dataclass
+class LaunchNodeError(CloudNodeProviderError):
+    # The node type that failed to launch.
+    node_type: NodeType
+    # Number of nodes that failed to launch.
+    count: int
+    # From which update request the error originates.
+    update_id: str
+
+
+@dataclass
+class TerminateNodeError(CloudNodeProviderError):
+    # The cloud instance id of the node that failed to terminate.
+    cloud_instance_id: CloudInstanceId
+    # From which update request the error originates.
+    update_id: str
+
+
+@dataclass
+class CloudNodeProviderState:
+    """
+    The state of a cloud node provider.
+    """
+
+    # The cloud nodes that are currently running.
+    running: Dict[CloudInstanceId, CloudInstance] = field(default_factory=dict)
+    # Errors that have happened when launching nodes.
+    launch_errors: List[LaunchNodeError] = field(default_factory=list)
+    # Errors that have happened when terminating nodes.
+    termination_errors: List[TerminateNodeError] = field(default_factory=list)
+
+
+class ICloudNodeProvider(ABC):
```

Comment:
nit: I feel it can just be named as `NodeProvider`

Reply:
I think cloud node or cloud instance is better given the concept of "node" is a bit ambiguous given we also call ray node a node sometimes. So I would think keeping "cloud" in the name is better, and given we are already calling cloud instance in many other places, I feel this might just be easier and more consistent.